File: examples/python/list_decks.py
# SPDX-FileCopyrightText: 2026 Adam Fidel
# SPDX-License-Identifier: MIT

"""Every deck installed on this system.

    python list_decks.py                # the XDG deck library
    python list_decks.py ~/decks        # a library root of your own
"""

import sys

import arcana_tarot as arcana


def main(argv: list[str]) -> int:
    # An empty `roots` means the XDG library, $XDG_DATA_HOME/tarot/decks.
    roots = [argv[1]] if len(argv) > 1 else []

    print("searching", arcana.deck_library_path())

    # Scanning happens in the constructor: it reads each deck.toml, and nothing
    # else. No images are touched.
    library = arcana.deck_library(arcana.library_options(roots=roots))

    for root in library.roots():
        print("  root", root)

    # Sorted by directory name. Roots are searched in order like PATH, so a
    # directory name found in an earlier root shadows a later one.
    for deck in library.decks():
        print(f"{deck.directory_name:24} {deck.name:32} v{deck.version:8} {deck.card_count:3} cards")
        print(f"{'':24} id={deck.id} at {deck.path}")

    # A directory whose manifest could not be read is reported rather than
    # dropped, so a broken deck is visible to the user instead of missing.
    for deck in library.malformed_decks():
        print(f"malformed {deck.directory_name}: {deck.problem.message}")

    print(f"{len(library.decks())} readable, {len(library.malformed_decks())} malformed")
    return 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))

File: examples/list_decks.cpp
// SPDX-FileCopyrightText: 2026 Adam Fidel
// SPDX-License-Identifier: MIT

// Every deck installed on this system.
//
//   ./list_decks                     # the XDG deck library
//   ./list_decks ~/decks             # a library root of your own

#include <arcana/library.hpp>
#include <arcana/paths.hpp>

#include <print>
#include <string_view>
#include <vector>

int main(int argc, char** argv)
{
    // empty root uses $XDG_DATA_HOME/tarot/decks
    arcana::library_options options;
    if (argc > 1)
        options.roots.emplace_back(argv[1]);

    std::println("searching {}", arcana::paths::deck_library_path().string());

    arcana::deck_library library{options};

    for (std::filesystem::path const& root : library.roots())
        std::println("  root {}", root.string());

    // Print all decks
    for (arcana::deck_summary const& deck : library.decks())
    {
        std::println(
            "{:24} {:32} v{:8} {:3} cards", deck.directory_name, deck.name, deck.version,
            deck.card_count
        );
        std::println(
            "{:24} id={} at {}", "", deck.identifier.value_or("(unset)"), deck.path.string()
        );
    }

    // Report broken decks
    for (arcana::malformed_deck const& deck : library.malformed_decks())
        std::println("malformed {}: {}", deck.directory_name, deck.problem.message);

    std::println(
        "{} readable, {} malformed", library.decks().size(), library.malformed_decks().size()
    );

    return 0;
}
